Add way to extract result/exceptions from thrift calls more easily

phile314 opened the pull request:
This class is mostly useful together with the code generator, as it will generate the class instances for the result datatypes automatically.

The class is in the RPC module as I plan to add another function which goes the opposite direction (SomeException -> Result datatype) for the server side.

phile314 commented on src/Pinch/Internal/RPC.hs:
```diff
@@ -40,3 +48,7 @@ readMessage chan = Transport.readMessage (cTransportIn chan) $ deserializeMessag
+
+class (Pinchable a, Tag a ~ TStruct) => ThriftResult a where
+  type ResultType a
+  toEither :: a -> Either SomeException (ResultType a)
```
Example code;

```
data Success_Result
  = Success_Result_Success Data.Text.Text
  deriving (Prelude.Eq, GHC.Generics.Generic, Prelude.Show)

instance Pinch.Pinchable Success_Result where
  .....

instance Pinch.ThriftResult Success_Result where
  type (ResultType Success_Result) = Data.Text.Text
  toEither (Success_Result_Success x) = Prelude.Right (x)
```

phile314 replied:
Hm, on second thoughts, this could also be encoded more directly:

```
class .... ThriftResult a where
  extract :: a -> IO (ResultType a)

instance Pinch.ThriftResult Example_Result where
  type (ResultType Example_Result) = Data.Text.Text
  extract (Example_Result_Success x) = pure x
  extract (Example_Result_MyError e) = throwIO e
```

What do you think?

abhinav replied:
I like the exception version since that's how we're surfacing failures in the
rest of the client/server logic.

FWIW, that approximately matches the helper we generate for this purpose in
our Go variant: a `UnwrapResponse` function that accepts the `*_Result` type
and returns either the return value of that function, or the exception as an
error.

abhinav reviewed (approved):
I'm in favor of the `extract :: a -> IO (ResultType a)`
approach unless there's a good reason to stick with `Either
SomeException a`.

If we were willing to take on more dependencies, we could
use `MonadThrow` and let the caller decide whether they
wanted `IO Result` or `Either SomeException Result` but I
don't know if there's need of that yet. I'm willing to stick
fine with one of the existing approaches you're considering
(concrete `IO` or `Either`).

phile314 commented:
I changed it to use the `a -> IO (ResultType a)` approach. I quite like the name `unwrap`, so I also changed the name to that ;-)

I intend to directly use `IO` for now, but yes, if it's useful for anybody I would have no issue with using MonadThrow.
